### PR TITLE
Make leaf differentiable and vectorized

### PR DESCRIPTION
**Pull request by SCiarella**
This PR addresses #45.

It follows the test structure of #55 to ensure that leaf parameters are differentiable and vectorizable.
Before merging, the issues with the vectorization of `drv` should be addressed.

### CODE REVIEW

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 1]
```diff
@@ -172,17 +186,18 @@ def initialize(
         DVS = self.kiosk["DVS"]
 
         params = self.params
-        shape = _get_params_shape(params)
+        params.shape = _get_params_shape(params)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/leaf_dynamics.py]:
```suggestion
        self.params_shape = _get_params_shape(params)
```

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 2]
```diff
@@ -172,17 +186,18 @@ def initialize(
...
+        DWLV = torch.zeros(params.shape, dtype=DTYPE)
         TWLV = WLV + DWLV
 
         # Initialize leaf classes (SLA, age and weight)
-        SLA = torch.zeros((*shape, self.MAX_DAYS), dtype=DTYPE)
-        LVAGE = torch.zeros((*shape, self.MAX_DAYS), dtype=DTYPE)
-        LV = torch.zeros((*shape, self.MAX_DAYS), dtype=DTYPE)
+        SLA = torch.zeros((*params.shape, self.MAX_DAYS), dtype=DTYPE)
+        LVAGE = torch.zeros((*params.shape, self.MAX_DAYS), dtype=DTYPE)
+        LV = torch.zeros((*params.shape, self.MAX_DAYS), dtype=DTYPE)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/leaf_dynamics.py]: can you place replace `params.shape` with `self.params_shape`?

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 3]
```diff
@@ -172,17 +186,18 @@ def initialize(
         DVS = self.kiosk["DVS"]
 
         params = self.params
-        shape = _get_params_shape(params)
+        params.shape = _get_params_shape(params)
+        self._shape_finalized = False  # Track if shape has been updated with drv data
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py]: please see my comment [here](https://github.com/WUR-AI/diffWOFOST/pull/56/files#r2541109312).

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@SCiarella thanks for implementation and tests 👍 . Please have a look at my comments. The main issue with the current implementation is that the physical models shouldn’t be worry about the shape or format of their inputs. At this stage, we assume the physical model receives correctly shaped data, so it can do vectorized computations and proper gradient tracking. Everything else should be handled by the Engine. The physical model can include basic checks, but it shouldn’t go more than that. 

If anything in my comments is unclear, please let me know, we can discuss the details.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 4]
```diff
@@ -232,6 +291,9 @@ def calc_rates(self, day: datetime.date, drv: WeatherDataContainer) -> None:
         p = self.params
         k = self.kiosk
 
+        # Update shape if needed based on drv.TEMP
+        self._ensure_shape_finalized(drv)
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py]: Ok I removed this function. However now if the user passes a scalar parameters while `drv.TEMP` is a tensor, this throws an error instead of automatically reshaping the parameters to accommodate that. If we do not want to use a function like `_check_drv_shape` but still support such a situation, then we would have to change the init method of the `WOFOST_Leaf_Dynamics`.

File: src/diffwofost/physical_models/utils.py [hunk 5]
```diff
@@ -556,6 +564,73 @@ def _get_params_shape(params):
...
+            # Scalar is valid, will be broadcast
+            continue
+        elif var_value.shape == expected_shape:
+            # Matches expected shape
+            continue
+        else:
+            raise ValueError(
+                f"Weather variable '{var_name}' has incompatible shape {var_value.shape}. "
+                f"Expected scalar (0-dimensional) or shape {expected_shape}."
+            )
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/diffwofost/physical_models/utils.py]: Very good suggestion! I have reduced the scope of this function to process single `drv.variables`. It is now called `_get_drv`.
